### ISSUE TITLE
HttpClientHandler_SslProtocols_Test.GetAsync_AllowedSSLVersion_Succeeds test failed

### ISSUE THREAD

**Issue by stephentoub** [dotnet/corefx#13281]
HttpClientHandler_SslProtocols_Test.GetAsync_AllowedSSLVersion_Succeeds test failed

https://ci.dot.net/job/dotnet_corefx/job/master/job/outerloop_windows_nt_debug_prtest/93/consoleText
```
System.Net.Http.Functional.Tests.HttpClientHandler_SslProtocols_Test.GetAsync_AllowedSSLVersion_Succeeds(acceptedProtocol: Tls11, requestOnlyThisProtocol: False) [FAIL]
        System.Net.Http.HttpRequestException : An error occurred while sending the request.
        ---- System.Net.Http.WinHttpException : The operation has been canceled
        Stack Trace:
              at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
              at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
              at System.Net.Http.HttpClient.<FinishSendAsync>d__58.MoveNext()
           --- End of stack trace from previous location where exception was thrown ---
              at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
           D:\j\workspace\outerloop_win---7ad0ce03\src\System.Net.Http\tests\FunctionalTests\HttpClientHandlerTest.SslProtocols.cs(102,0): at System.Net.Http.Functional.Tests.HttpClientHandler_SslProtocols_Test.<>c__DisplayClass4_0.<<GetAsync_AllowedSSLVersion_Succeeds>b__0>d.MoveNext()
           --- End of stack trace from previous location where exception was thrown ---
              at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
           D:\j\workspace\outerloop_win---7ad0ce03\src\Common\tests\System\Net\Http\LoopbackServer.cs(57,0): at System.Net.Test.Common.LoopbackServer.<>c__DisplayClass3_0.<CreateServerAsync>b__0(Task t)
              at System.Threading.Tasks.Task.Execute()
           --- End of stack trace from previous location where exception was thrown ---
              at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
           D:\j\workspace\outerloop_win---7ad0ce03\src\System.Net.Http\tests\FunctionalTests\HttpClientHandlerTest.SslProtocols.cs(100,0): at System.Net.Http.Functional.Tests.HttpClientHandler_SslProtocols_Test.<GetAsync_AllowedSSLVersion_Succeeds>d__4.MoveNext()
           --- End of stack trace from previous location where exception was thrown ---
              at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           --- End of stack trace from previous location where exception was thrown ---
              at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           --- End of stack trace from previous location where exception was thrown ---
              at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           ----- Inner Stack Trace -----
              at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
              at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
              at System.Net.Http.WinHttpHandler.<StartRequest>d__105.MoveNext()
```

**Comment by danmosemsft** [comment]
@davidsh @CIPop do you want to investigate or disable this one?

**Comment by davidsh** [comment]
We will probably just need to disable it. But in reality, the cause is the same:

```
 System.Net.Http.HttpRequestException : An error occurred while sending the request.
        ---- System.Net.Http.WinHttpException : The operation has been canceled
```

That basically means that there was a connection error of some sort (timeout causing cancellation) connecting to the Azure test endpoint.  This is why we are trying to mitigate some of this by moving more tests to LoopBack server.  But doing SSL testing with loopback can be difficult.

**Comment by stephentoub** [comment]
This test is  actually already using a loopback server. So if it's timing out after a minute, to me that suggests an actual bug somewhere, either src or test.

**Comment by davidsh** [comment]
It might not be a bug in the test.  It is possible that if multiple PR's are validating on the same machine (physical or VM), then ephemeral ports might run out.  Or, the CPU itself is pegged to 100% doing stuff (compiling, run tests etc.) and thus the test code could timeout as well.

**Comment by Jiayili1** [comment]
This issue is repro on OuterLoop_Win10_debug ([build#108](https://ci.dot.net/job/dotnet_corefx/job/master/job/outerloop_win10_release/108/testReport/))

Detail: https://ci.dot.net/job/dotnet_corefx/job/master/job/outerloop_win10_release/108/testReport/System.Net.Http.Functional.Tests/HttpClientHandler_SslProtocols_Test/GetAsync_AllowedSSLVersion_Succeeds_acceptedProtocol__Tls__requestOnlyThisProtocol__True_/

Message:
~~~
System.Net.Http.HttpRequestException : An error occurred while sending the request.\r\n---- System.Net.Http.WinHttpException : The operation has been canceled
~~~

Stack Trace:
~~~
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at System.Net.Http.HttpClient.<FinishSendAsyncBuffered>d__58.MoveNext() in D:\j\workspace\outerloop_win---6835b088\src\System.Net.Http\src\System\Net\Http\HttpClient.cs:line 462 --- End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at System.Net.Http.Functional.Tests.HttpClientHandler_SslProtocols_Test.<>c__DisplayClass4_0.<<GetAsync_AllowedSSLVersion_Succeeds>b__0>d.MoveNext() in D:\j\workspace\outerloop_win---6835b088\src\System.Net.Http\tests\FunctionalTests\HttpClientHandlerTest.SslProtocols.cs:line 102 --- End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at System.Net.Test.Common.LoopbackServer.<>c__DisplayClass3_0.<CreateServerAsync>b__0(Task t) in D:\j\workspace\outerloop_win---6835b088\src\Common\tests\System\Net\Http\LoopbackServer.cs:line 58 at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state) at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot) --- End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at System.Net.Http.Functional.Tests.HttpClientHandler_SslProtocols_Test.<GetAsync_AllowedSSLVersion_Succeeds>d__4.MoveNext() in D:\j\workspace\outerloop_win---6835b088\src\System.Net.Http\tests\FunctionalTests\HttpClientHandlerTest.SslProtocols.cs:line 100 --- End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) --- End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) --- End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) ----- Inner Stack Trace ----- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Threading.Tasks.RendezvousAwaitable`1.GetResult() in D:\j\workspace\outerloop_win---6835b088\src\Common\src\System\Threading\Tasks\RendezvousAwaitable.cs:line 62 at System.Net.Http.WinHttpHandler.<StartRequest>d__105.MoveNext() in D:\j\workspace\outerloop_win---6835b088\src\System.Net.Http.WinHttpHandler\src\System\Net\Http\WinHttpHandler.cs:line 851
~~~

**Comment by ianhays** [comment]
This failure may be related to #16516 and https://github.com/dotnet/corefx/issues/16534.

**Comment by stephentoub** [comment]
Again here:
https://ci.dot.net/job/dotnet_corefx/job/master/job/outerloop_windows_nt_debug_prtest/124/consoleText

**Comment by Priya91** [comment]
With the last observed failure to this one, seems like the failure is happening when winhttphandler is using multiple TLS protocols and the server supports only a single protocol. 

In the case when client uses multiple protocols that are not used by the server, the client handle state is not updated by the server, and it results in indefinite wait, until the default httpclient timeout happens. This is tracked by issue #8538. This issue repros locally consistently.

In the case of the current issue, seems like indefinite wait happens intermittently when client uses multiple protocols out of which one is supported by server. I was not able to repro this locally.

This looks like a bug in the native winhttphandler, and requires deeper investigation.

**Comment by davidsh** [comment]
I am closing this issue because we don't have a repro and there is no actionable information to diagnose.

**Comment by karelz** [comment]
We have record of 3 occurrences (~1 per month), I would really like to see us to do something about the test -- e.g. add logging.
It seems to be something test specific given that this is unique failure for this test.

I am fine pushing it to Future based on the cadence, but IMO we should not close it yet. Let's discuss it tomorrow in Networking sync up ...

**Comment by karelz** [comment]
Also I would like to understand how much we tried to repro.

For example, I would love to see one CI machine (same config) to be hammer with just networking tests running in a loop for a week. @steveharter is it something you considered?

**Comment by CIPop** [comment]
> I am closing this issue because we don't have a repro and there is no actionable information to diagnose.

I don't agree: if we don't have a repro and no actionable information it only means __we need__ to get a repro and get actionable information, not that we can push this under the rug and disable the test.

The reason we don't have actionable information is still having zero logs from CI: https://github.com/dotnet/corefx/issues/4152 . 
I will take this offline with @karelz and @steveharter to figure out a way to enable logs and automating crash-dumps generation using [MiniDumpWriteDump](https://msdn.microsoft.com/en-us/library/windows/desktop/ms680360(v=vs.85).aspx).

**Comment by karelz** [comment]
BTW: crash dumps are collected in CI as of recently. ETW traces will be next - I expect that issues like these and our attempts to make them actionable will drive that effort.

**Comment by CIPop** [comment]
> crash dumps are collected in CI as of recently

Just saw this: it is really great! 
We now need to change the code to force a crash when this happens, enable the test (if not done already) and look at the captured state.

**Comment by davidsh** [comment]
>BTW: crash dumps are collected in CI as of recently.

Is this documented somewhere, especially where we can find the dumps?

**Comment by karelz** [comment]
@mellinoe @sepidehMS @danmosemsft can you please point us to crash dump collection docs?

**Comment by karelz** [comment]
@CIPop or instead of waiting for next CI failure, we can leverage @steveharter's networking-test-suite-running-in-a-loop effort and hopefully get the actionable data in days.
I hope that early next week we will know which of the failures we are able to hit in such a system and they become suddenly actionable :) ... we still might want to push them out into Future based on CI impact (assuming we suspect test/infra bugs and not product stress issues).

**Comment by steveharter** [comment]
Yes I took a couple CI machines offline and will test some of the networking test assemblies repeatedly over a few days and see what that produces; capturing output for the failures. This will at least help determine the failure frequency for now. I've already been doing this for specific failures\issues, but now running it against all tests.

Once that is starts producing output, I can expand it to force crash for certain tests, or perhaps attempt to use Debugger.Break.

**Comment by karelz** [comment]
Cool, thanks a lot @steveharter!

**Comment by steveharter** [comment]
On the Win7 CI machine there was ~3% fail rate due to this error (15 failures out of 560 runs). No failures yet on Win10 CI machine.

I did not yet enable additional diagnostics for this error. PM me if you want access to these high-level results.

Note that although CI calls these "Win7" the real windows version is "Windows Server 2008 R2 Datacenter" with SP1.

**Comment by karelz** [comment]
Moving to Future due to low frequency (1 per month in Jenkins CI).

**Comment by CIPop** [comment]
Closed as dupe of #19701
